### PR TITLE
Reroll initiative refactor - ver 0.1

### PR DESCRIPTION
**Pull request by eclarke12**
Back-to-basics refactor of the module.

### CODE REVIEW

File: reroll-initiative/old/reroll-initiative.js [hunk 1]
```diff
@@ -0,0 +1,260 @@
...
+        this.config = RerollIinitiative.DEFAULT_CONFIG;
+        console.log("Resetting reroll-initiative settings to defaults:",RerollInitiative.DEFAULT_CONFIG);
+        this._saveSettings();
+    }
+
+    /**
+     * Saves current class instance options back to game settings
+     */
+    async _saveSettings () {
+        await game.settings.set(RerollInitiative.SETTINGS.module,RerollInitiative.SETTINGS.key,this.config);
```
Comment (Rawrly85) [inline reroll-initiative/old/reroll-initiative.js]: Minor: I always like from a readability aspect to new line arguments when there are more than 2. It's easier to scan the code and understand. Completely a preference thing.

File: reroll-initiative/old/reroll-initiative.js [hunk 2]
```diff
@@ -0,0 +1,260 @@
...
+    }
+
+    /**
+     * Loads current class instance settings from game settings
+     */
+    async _loadSettings () {
+        let config = await game.settings.get(RerollInitiative.SETTINGS.module,RerollInitiative.SETTINGS.key);
```
Comment (Rawrly85) [inline reroll-initiative/old/reroll-initiative.js]: No need for the local var, just directly assigned it

File: reroll-initiative/old/reroll-initiative.js [hunk 3]
```diff
@@ -0,0 +1,260 @@
...
+    static async updateSetting(setting,value){
+        let settings = game.settings.get(RerollInitiative.SETTINGS.module,RerollInitiative.SETTINGS.key);
+
+        if(settings.hasOwnProperty(setting)){
+            console.log(setting);
+            settings[setting] = value;
+            await game.settings.set(RerollInitiative.SETTINGS.module,RerollInitiative.SETTINGS.key,settings);
+        }
+        else{
+            console.error("Module setting : "+setting+" does not exist!");
```
Comment (Rawrly85) [inline reroll-initiative/old/reroll-initiative.js]: Is the pattern to console.error and not actually throw an error?

File: reroll-initiative/old/reroll-initiative.js [hunk 4]
```diff
@@ -0,0 +1,260 @@
...
+    static querySettings() {
+        return game.settings.get(RerollInitiative.SETTINGS.module,RerollInitiative.SETTINGS.key)
+    }
+
+    /**
+     * Hook on combat update and if round in update is greater than previous -- call resetAndReroll
+     */
+    _postUpdateCombatHook() {
+        Hooks.on("updateCombat", (combat,update) =>  {
+            //this is probably an unnecessary hit on the db
```
Comment (Rawrly85) [inline reroll-initiative/old/reroll-initiative.js]: What's the plan for all the commented out code?

File: reroll-initiative/old/reroll-initiative.js [hunk 5]
```diff
@@ -0,0 +1,260 @@
...
+
+            if(html){
+                let rriCheckbox = this._injectCheckboxFormgroup(html);
+
+                if(rriCheckbox){
+                    //Set the state of the checkbox to match the current module reroll setting
+                    rriCheckbox.prop("checked",settings.reroll);
+
+                    // Adjust the window height to allow for new content
+                    app.setPosition({height: app.position.height + 60});
```
Comment (Rawrly85) [inline reroll-initiative/old/reroll-initiative.js]: Why 60? Does this work on different resolutions?

File: reroll-initiative/old/reroll-initiative.js [hunk 6]
```diff
@@ -0,0 +1,260 @@
...
+                
+                    // Handle form submission
+                    //todo: break this out into a separate method?
+                    const submit = html.find('button[type="submit"]');
```
Comment (Rawrly85) [inline reroll-initiative/old/reroll-initiative.js]: My previously comment is still valid here. Finding by this is usually to generic. If there are multiple submit buttons on the page, you could potentially grab the wrong one. You should look at giving your button a unique ID and doing he lookup via that id

File: reroll-initiative/old/reroll-initiative.js [hunk 7]
```diff
@@ -0,0 +1,260 @@
...
+        });
+    }
+
+    /**
+     * Injects a checkbox formgroup inside the designated element
+     * @param {*} html
+     * @returns {Object} checkbox The checkbox element that was injected
+     */
+    _injectCheckboxFormgroup(html){
+        const elementIdentifier = 'button[type="submit"]';
```
Comment (Rawrly85) [inline reroll-initiative/old/reroll-initiative.js]: This identifier is used above, you'll probably want to pull this constant to the other constant file and use it in both cases (unless you solve my question above with the unique id and no longer need this)

File: reroll-initiative/old/reroll-initiative.js [hunk 8]
```diff
@@ -0,0 +1,260 @@
...
+                <div class="form-group">
+                    <label>${label}</label>
+                    <input type="checkbox" name=${name} data-dtype="Boolean">
+                    <p class=hint>${hint}</p>
+                </div>`
+            );
+            console.log(html);
+        }
+        else {
+            console.log("Failed to inject checkbox... " +elementIdentifier+ " does not exist.")
```
Comment (Rawrly85) [inline reroll-initiative/old/reroll-initiative.js]: Throw error?

File: reroll-initiative/old/reroll-initiative.js [hunk 9]
```diff
@@ -0,0 +1,260 @@
...
+                }
+            }
+        });
+    }
+
+    /**
+     * Injects a checkbox formgroup inside the designated element
+     * @param {*} html
+     * @returns {Object} checkbox The checkbox element that was injected
+     */
```
Comment (Rawrly85) [inline reroll-initiative/old/reroll-initiative.js]: This function is doing to much. Id separate the logic to inject and the logic to return.

File: reroll-initiative/reroll-initiative.js [hunk 10]
```diff
@@ -0,0 +1,126 @@
...
+    rerollInitiative();
+});
+
+/**
+ * Main module function
+ */
+function rerollInitiative() {
+    console.log("reroll-initiative starting up...")
+    
+    const MODULE_NAME = "reroll-initiative";
```
Comment (Rawrly85) [inline reroll-initiative/reroll-initiative.js:19]: You define this constants multiple times for the same string. Why not have a class that has static methods to return this string names instead or a class with all these constants you can access?

File: reroll-initiative/old/reroll-initiative.js (hunk 2 above)
Comment (eclarke12) [reply reroll-initiative/old/reroll-initiative.js]: Oh yeah not sure how that one slipped in. I think I was trying to work around a different issue. I will remove the declaration.

File: reroll-initiative/old/reroll-initiative.js (hunk 3 above)
Comment (eclarke12) [reply reroll-initiative/old/reroll-initiative.js]: It should probably be a try...catch block. Looking here: https://stackoverflow.com/questions/25377115/what-is-the-difference-between-throw-error-and-console-error I should probably confirm what behaviour is required.

**Review by glimmbo — DISMISSED** [review]
Still not 100% following everything in here, but I'll say "Looks good!"

File: reroll-initiative/reroll-initiative.js (hunk 10 above)
Comment (eclarke12) [reply reroll-initiative/reroll-initiative.js:19]: Can you please clarify this comment?
Regarding classes: this design pattern is meant to be classless and simple.

Comment (Rawrly85) [reply reroll-initiative/reroll-initiative.js:19]: After the latest change, this comment is no longer relevant.

File: reroll-initiative/reroll-initiative.js [hunk 11]
```diff
@@ -0,0 +1,126 @@
...
+            if(settings[SETTING]) {
+                return "checked"
+            }
+            else {
+                return ""
+            }
+        }
+
+        //find the submit button on the form then inject the new checkbox formgroup above it
+        const submit = html.find('button[type="submit"]');
```
Comment (Rawrly85) [inline reroll-initiative/reroll-initiative.js:112]: I made a comment in the previous deleted file but this sort of code can be dangerous as the find isn't looking up something uniquely. I'd look at using an ID instead.

File: reroll-initiative/reroll-initiative.js [hunk 12]
```diff
@@ -0,0 +1,126 @@
...
+            settings = s;
+            console.log("Settings changed to",s);
+        }
+    }
+
+    //initialise an object to hold settings in memory. assign the default config for now.
+    let settings = DEFAULT_CONFIG;
+
+    //register settings with the game
+    game.settings.register(MODULE_NAME,SETTINGS_NAME,SETTINGS_META);
```
Comment (Rawrly85) [inline reroll-initiative/reroll-initiative.js]: Minor: this is hard to read. 

My preference is to new line each argument once there are more than 2. At the very least, spaces between should be there.

File: reroll-initiative/reroll-initiative.js [hunk 13]
```diff
@@ -0,0 +1,126 @@
...
+ */
+Hooks.on("ready", () => {
+    rerollInitiative();
+});
+
+/**
+ * Main module function
+ */
+function rerollInitiative() {
+    console.log("reroll-initiative starting up...")
```
Comment (Rawrly85) [inline reroll-initiative/reroll-initiative.js]: Are all these comment logs specifically used for your own debugging purpose? Id probably remove them from production as the end user probably doesn't care/can't do much about it.

File: reroll-initiative/reroll-initiative.js [hunk 14]
```diff
@@ -0,0 +1,126 @@
...
+    const MODULE_NAME = "reroll-initiative";
+
+    const DEFAULT_CONFIG = {
+        reroll: true
+    };
+
+    const SETTINGS_NAME = "rriSettings";
+
+    const SETTINGS_META = {
+        name: "rriSettings",
```
Comment (Rawrly85) [inline reroll-initiative/reroll-initiative.js]: Use the constant defined above

Comment (eclarke12) [reply reroll-initiative/reroll-initiative.js]: duplicate/typo. will review and remove the extra

File: reroll-initiative/reroll-initiative.js (hunk 11 above)
Comment (eclarke12) [reply reroll-initiative/reroll-initiative.js:112]: agreed. currently with the tools provided it is not within scope to have a more unique lookup. noted for future coding that ids will be targeted first, then name, then type/other props last.

File: reroll-initiative/reroll-initiative.js (hunk 12 above)
Comment (eclarke12) [reply reroll-initiative/reroll-initiative.js]: i will add spaces.

File: reroll-initiative/reroll-initiative.js (hunk 13 above)
Comment (eclarke12) [reply reroll-initiative/reroll-initiative.js]: provides a paper trail for enduser troubleshooting.

**Review by glimmbo — DISMISSED** [review]

**Review by Rawrly85 — DISMISSED** [review]
Given the lack of proper ID reference from Foundry itself for access Dom elements, I'm okay with the review. I'd highly suggest creating a request for the foundry people to add IDs to elements or at least return a scoped version of the HTML to remove the problemaic issues that can occur writing code this way.

**Review by glimmbo — DISMISSED** [review]

**Review by glimmbo — APPROVED** [review]

**Pull request closed without merging**